libraries/logging/README.md

```diff
@@ -0,0 +1,109 @@
+## Sample Implementation of Logging Interface
```

Comment:
## Sample Implementation of Logging Interface

This library provides reference implementations of the
[logging interface](#the-logging-interface) that is used by FreeRTOS libraries
and demos to generate log messages. This library provides implementations for
both ISO C90 standard and ISO C99 standard (with GNU extension) of the logging
interface.

**Note**: Users can provide their own implementations of the logging interface
to enable logging, if they choose to not use this implementation.

To enable logging using the sample implementation for a FreeRTOS library and/or
demo:

1. Configure logging level by defining `LIBRARY_LOG_LEVEL`  macro to one of
`LOG_NONE`, `LOG_ERROR`, `LOG_WARN`, `LOG_INFO`, `LOG_DEBUG`.
1. If using the ISO C99 (with GNU extension) implementation (which is default),
define `LIBRARY_LOG_NAME` to configure the library name which is prepended to
the log messages.
1. If your compiler only supports the ISO C90 standard, use the C90
implementation by setting the `LOGGING_ENABLE_METADATA_WITH_C99_AND_GNU_EXTENSION`
macro to `0`.

The logging file include and macro definitions must follow the following order:
1. Include `logging_levels.h`.
2. Define  `LIBRARY_LOG_LEVEL`, `LIBRARY_LOG_NAME`, and
`LOGGING_ENABLE_METADATA_WITH_C99_AND_GNU_EXTENSION` as mentioned above.
3. Include `logging_stack.h`.

A sample logging configuration for the coreMQTT library which is defined in
`core_mqtt_config.h` file:

```c
#include "logging_levels.h"

#ifndef LIBRARY_LOG_NAME
    #define LIBRARY_LOG_NAME    "MQTT"
#endif

#ifndef LIBRARY_LOG_LEVEL
    #define LIBRARY_LOG_LEVEL    LOG_INFO
#endif

#include "logging_stack.h"
```

The above configuration results in the logging messages looking like the following.
Note that the library name (`[MQTT]`) and source file location (`[core_mqtt.c:1175]`)
are prepended to each log message.
```
78 1950 [iot_thread] [INFO] [MQTT] [core_mqtt.c:886] Packet received. ReceivedBytes=2.
79 1950 [iot_thread] [INFO] [MQTT] [core_mqtt.c:1162] Ack packet deserialized with result: MQTTSuccess.
80 1951 [iot_thread] [INFO] [MQTT] [core_mqtt.c:1175] State record updated. New state=MQTTPublishDone.
```

If ISO C90 implementation is used (i.e. `LOGGING_ENABLE_METADATA_WITH_C99_AND_GNU_EXTENSION`
is defined to `0`), the logging messages look like the following. Note that the
library name and source file location are not present in each log message.
```
78 1950 [iot_thread] [INFO] Packet received. ReceivedBytes=2.
79 1950 [iot_thread] [INFO] Ack packet deserialized with result: MQTTSuccess.
80 1951 [iot_thread] [INFO] State record updated. New state=MQTTPublishDone.
```

## Using your custom implementation of Logging Interface

The Logging Interface comprises of the following 4 logging macros, listed in
increasing order of verbosity:

1. `LogError`
2. `LogWarn`
3. `LogInfo`
4. `LogDebug`

`LogError` is only called when there is an error and so is the least verbose,
whereas `LogDebug` is called more frequently to provide debug level information
and is therefore, the most verbose.

An implementation of the logging interface needs to define these macros to
map them to their logging implementation. Logging macros use format specifier
and variable number of arguments, just like standard library `printf` function
but they use double parenthesis to support ISO C90 which should be taken care
while defining them.

An example call of the logging macros is the following:

```
LogInfo( ( "This prints an integer %d", 100 ) );
```

If you have a thread safe `printf` function, `LogInfo` should be defined like
the following. Notice how one set of parentheses are removed because `X`
already contains a set of parentheses.

```
#define LogInfo( X )  printf X
```

Reply:
Thank you for providing a concise version of the README.